packages/flutter_tools/lib/src/flutter_plugins.dart

```diff
@@ -979,3 +1048,178 @@ Future<void> injectPlugins(
+                '\n\n'
+              );
+            }
+            didFindError = true;
+          }
+          // Use the plugin implementation added by the user as a direct dependency.
+          continue;
+        }
+      }
+      directDependencyResolutions[resolutionKey] = PluginInterfaceResolution(
+        plugin: plugin,
+        platform: platform,
+      );
+    }
+  }
+  if (didFindError && throwOnPluginPubspecError) {
+    throwToolExit('Please resolve the errors');
+  }
+  final List<PluginInterfaceResolution> finalResolution = <PluginInterfaceResolution>[];
+  for (final MapEntry<String, PluginInterfaceResolution> resolution in directDependencyResolutions.entries) {
+    if (resolution.value.plugin.isDirectDependency) {
+      finalResolution.add(resolution.value);
+    } else if (defaultImplementations.containsKey(resolution.key)) {
+      // Pick the default implementation.
+      if (defaultImplementations[resolution.key] == resolution.value.plugin.name) {
+        finalResolution.add(resolution.value);
+      }
+    }
+  }
+  return finalResolution;
+}
+
+/// Generates the Dart plugin registrant, which allows to bind a platform
+/// implementation of a Dart only plugin to its interface.
+/// The new entrypoint wraps [currentMainUri], adds the [_PluginRegistrant] class,
+/// and writes the file to [newMainDart].
+///
+/// [mainFile] is the main entrypoint file. e.g. /<app>/lib/main.dart.
+///
+/// A successful run will create a new generate_main.dart file or update the existing file.
+/// Throws [ToolExit] if unable to generate the file.
+///
+/// This method also validates each plugin's pubspec.yaml, but errors are only
+/// reported if [throwOnPluginPubspecError] is [true].
+///
+/// For more details, see https://flutter.dev/go/federated-plugins.
+Future<void> generateMainDartWithPluginRegistrant(
+  FlutterProject rootProject,
+  PackageConfig packageConfig,
+  String currentMainUri,
+  File newMainDart,
+  File mainFile, {
+  bool throwOnPluginPubspecError,
+}) async {
+  final List<Plugin> plugins = await findPlugins(rootProject);
+  final List<PluginInterfaceResolution> resolutions = resolvePlatformImplementation(
+    plugins,
+    throwOnPluginPubspecError: throwOnPluginPubspecError ?? false,
+  );
+  final LanguageVersion entrypointVersion = determineLanguageVersion(
+    mainFile,
+    packageConfig.packageOf(mainFile.absolute.uri),
+    Cache.flutterRoot,
+  );
+  final Map<String, dynamic> templateContext = <String, dynamic>{
+    'mainEntrypoint': currentMainUri,
+    'dartLanguageVersion': entrypointVersion.toString(),
+    LinuxPlugin.kConfigKey: <dynamic>[],
+    MacOSPlugin.kConfigKey: <dynamic>[],
+    WindowsPlugin.kConfigKey: <dynamic>[],
+  };
+  if (resolutions.isEmpty) {
+    try {
+      if (newMainDart.existsSync()) {
+        newMainDart.deleteSync();
+      }
+    } on FileSystemException catch (error) {
+      throwToolExit(
```

Comment:
What sorts of fs exceptions are you expecting here. The tool already has some built in handling for permission/disk corruption issues. Other errors might be programming errors so I'd prefer if they made it to crash logging.

Could you update this to print the message as an error and then rethrow the original exception so we get information from crash logging?

Reply:
The documentation for the methods used above says that an exception may be thrown although I don't know the platforms, environments, and conditions in which that exception can be effectively thrown.